### PR TITLE
lammps: new stable versions, deprecate older patch versions

### PR DESCRIPTION
**Pull request by rbberger**
Add new stable versions:
- `lammps@20230802`
- `lammps@20220623.4`

Given that `20230802` supersedes features releases (aka patch versions), those are now all marked deprecated.
Kokkos 4 support has only been merged into `develop` and was not part of the stable release.

### CODE REVIEW

File: var/spack/repos/builtin/packages/lammps/package.py
Comment (alalazo) [inline var/spack/repos/builtin/packages/lammps/package.py:80]: This is both `preferred` and `deprecated`, can we remove the former?

**Review by alalazo — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/lammps/package.py
Comment (rbberger) [reply var/spack/repos/builtin/packages/lammps/package.py:80]: No, because it would change the hash. I'm using `preferred` to determine whether it's a stable or patch version, which changes the download URL.

Comment (rbberger) [reply var/spack/repos/builtin/packages/lammps/package.py:80]: Btw, if that hack is frowned upon, maybe the cleaner solution is to have a separate `stable_versions` list/set that I then use in `get_url_for_version`.

File: var/spack/repos/builtin/packages/lammps/package.py
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/lammps/package.py]: Confirmed this sha256 but is there a reason to not include `sha256=` for the argument (as is done with the other version directives)?

File: var/spack/repos/builtin/packages/lammps/package.py
Comment (alalazo) [reply var/spack/repos/builtin/packages/lammps/package.py:80]: > No, because it would change the hash.

Can you elaborate on that? If you're referring to the DAG hash, that shouldn't change if you remove `preferred=True`

Comment (rbberger) [reply var/spack/repos/builtin/packages/lammps/package.py:80]: The sha256 checksum changes since it downloads a different file if it's a patch or stable release (both files exist, since every stable also has a patch tarball)

File: var/spack/repos/builtin/packages/lammps/package.py
Comment (rbberger) [reply var/spack/repos/builtin/packages/lammps/package.py]: That's a typo, I'll fix it.

File: var/spack/repos/builtin/packages/lammps/package.py
Comment (alalazo) [reply var/spack/repos/builtin/packages/lammps/package.py:80]: I see the logic in `url_for_version`. It's not the prettiest, but let's leave it for another PR focused on that.

File: var/spack/repos/builtin/packages/lammps/package.py
Comment (rbberger) [reply var/spack/repos/builtin/packages/lammps/package.py]: @tldahlgren Fixed.

File: var/spack/repos/builtin/packages/lammps/package.py
Comment (rbberger) [reply var/spack/repos/builtin/packages/lammps/package.py:80]: @alalazo Ok, if you could then approve this one? Once merged, I'll create a new one that avoids this logic.

**Review by alalazo — APPROVED** [review]

**Pull request merged**
